File: 医检互认日志处理/analysis_log_files.py
# This is a sample Python script.
import json
import os
from itertools import filterfalse
import sys
from rich import print
from datetime import datetime
import sqlite3
from unit import FixedSizeDeque
from query_from_local_db import query_from_local_db,QueryType

# Press Shift+F10 to execute it or replace it with your code.
# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.



def calculate_age(birth_date_str: str):
    birth_date = datetime.strptime(birth_date_str, "%Y%m%d")
    today = datetime.today()
    print(today)
    print(birth_date)
    age = today.year - birth_date.year - ((today.month, today.day) < (birth_date.month, birth_date.day))
    return age


def get_file_item(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    patients_item = FixedSizeDeque()
    patients_list = []
    for line in lines:
        age = patient_name = patient_id = patient_birthday = patient_gender = patient_diagnose = patient_phone = patient_address = patient_name = patient_new_item = ""
        type_1 = False
        type_2 = False
        patients_item.add(line)
        if "要打开" in line:
            # print(file_path)
            patients_item_to_list = patients_item.get_all()
            entry_line = patients_item_to_list[0]
            leave_line = patients_item_to_list[1]
            log_date = entry_line.split("====>")[0]
            # print(entry_line)
            # print(leave_line)
            try:
                entry_info = json.loads(entry_line.split("入参:")[1])
                leave_info = json.loads(leave_line.split("出参:")[1])
            except:
                continue
            # print(leave_info)

            if "患者近期相似检查项目查询接口" in entry_line:
                type_1 = True
                print(entry_info)
                print(leave_info)
                patient_name = entry_info["patience"]["name"]
                patient_id = entry_info["patience"]["id_no"]
                patient_birthday = entry_info["patience"]["birthday"].replace("-","")
                patient_gender = "男" if entry_info['patience']['gender']== "1" else "女"
                patient_diagnose = query_from_local_db(patient_id,QueryType.DIAGNOSE)

                patient_phone = entry_info["patience"]["phone"]
                patient_address = entry_info["patience"]["address"]
                doctor_name = entry_info["doctor"]["name"]
                doctor_practice_number = entry_info["doctor"]["practice_number"]
                doctor_idcard = entry_info["doctor"]["idcard"]
                patient_new_item = [f'{item["project_name"]}:{item["study_body_part"]}'   for item in entry_info["visit_order"]["project_list"]]

                age = calculate_age(patient_birthday)
            if "您开立的项目有可互认的结果，请确认是否互认" in leave_line:
                type_2 = True
                # orign_info = json.loads(line.split("出参:")[1])
                patient_id = entry_info['messages']['PID']['identityNo']
                patient_name = entry_info['messages']['PID']['personalName']
                patient_gender = "男" if entry_info['messages']['PID']['sexCode'] == "1" else "女"
                patient_birthday = entry_info['messages']['PID']['birthDate']
                patient_diagnose = entry_info['messages']['DG1']['diagName']
                patient_new_item = entry_info['messages']['OBR']['lab'][0]["labItemName"]
                doctor_name = entry_info['messages']['PV1']['doctorName']
                patient_old_item = [i.split(' ')[0] for i in leave_info['data']['QAK']['brief']['entry']]
                # visit_date = entry_info['messages']['PV1']['visitDate']  # 这不是发生日期，发生日期还是得通过文件名。
                patient_phone = query_from_local_db(patient_id,QueryType.PHONE)
                patient_address = query_from_local_db(patient_id,QueryType.ADDRESS)
            if type_1 or type_2:
                # print(f"birthday:{patient_birthday}")
                age = calculate_age(patient_birthday)
                item_to_save = {
                            "patient_id": patient_id,
                            "name": patient_name,
                            "doctor": doctor_name,
                            "birthday": patient_birthday,
                            "gender": patient_gender,
                            "age": f"{age}岁",
                            "diagnose": patient_diagnose,
                            "patient_phone":patient_phone,
                            "patient_address":patient_address,
                            # "old_item": patient_old_item,
                            "new_item": patient_new_item,
                            # "visit_date": visit_date,
                            "log_date": log_date
                        }
                patients_list.append(item_to_save)
    print(patients_list)
    return patients_list


def filter_files(directory):
    items = os.listdir(directory)
    files = []
    for item in items:
        full_path = os.path.join(directory, item)
        if os.path.isfile(full_path):
            files.append(full_path)
    # print(files)
    print("files print end.")
    for file in files:
        patients_list = get_file_item(file)
        insert_db(patients_list)



def query_db(patient_item_to_save) -> bool:
    log_date = patient_item_to_save['log_date']
    sql = f"""select * from recognition where log_date = '{log_date}'"""
    result = cursor.execute(sql).fetchall()
    if len(result) < 1:
        return False
    return True


def insert_db(patient_item_to_save):
    sql = f"""insert into recognition values(?,?,?,?,?,?,?,?,?,?,?)"""
    if len(patient_item_to_save) < 1 or patient_item_to_save is None:
        return
    for item in patient_item_to_save:

        content = [item["log_date"], item["name"], item["patient_id"],
                   item["birthday"], item["gender"], item["age"],
                   item["diagnose"], item["patient_address"], "".join(item["new_item"]),
                   item["patient_phone"], item["doctor"]]
        if query_db(item):
            print("data already exists in recognition.")
            continue
        cursor.execute(sql, content)
        print("data insert success!")
    conn.commit()


def create_db():
    # 定义SQL语句来创建表
    create_table_sql = '''
    CREATE TABLE IF NOT EXISTS recognition (
        log_date TEXT PRIMARY KEY,
        name TEXT,
        patient_id TEXT,
        birthday TEXT,
        gender TEXT,
        age TEXT,
        diagnose TEXT,
        address TEXT,
        new_item TEXT,
        phone TEXT,
        doctor TEXT
    )
    '''
    conn.commit()
    # 执行SQL语句
    cursor.execute(create_table_sql)

    create_table_sql = '''
        CREATE TABLE IF NOT EXISTS zlbh_path (
            directory TEXT PRIMARY KEY
        )
        '''
    cursor.execute(create_table_sql)
    # 提交事务
    conn.commit()
    # 打印一个确认信息
    print("Table 'recognition' created successfully.")


# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    directory = "Log_医检互认"
    if len(sys.argv) > 1:
        directory = sys.argv[1]
        print(f"get cmd arg: {directory} !")
    else:
        print("no cmd arg!")

    conn = sqlite3.connect(r'D:\MytoolDataFiles\data\recognition_data.db')
    cursor = conn.cursor()
    create_db()
    filter_files(directory)
    conn.close()
#    · input(f"All log files in {directory} have been inserted into the database at D:\\MytoolDataFiles\\data\\recognition_data.db"
# )

File: MytoolMiniWPF/common/RecoginitionFunc/AnalysisLogFiles.cs
using MySql.Data.MySqlClient;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace MytoolMiniWPF.common
{
    
    internal class AnalysisLogFiles
    {
        string directory = @"D:\Log_医检互认";
        string connectionString = "server=localhost;port=3306;user=mytool;password=123321;database=mytooldata";

        public  AnalysisLogFiles(string dir,string connstr)
                {
                directory = dir;
                connectionString = connstr;
                }
        // 计算年龄
        public  int CalculateAge(string birthDateStr)
        {
            DateTime birthDate = DateTime.ParseExact(birthDateStr, "yyyyMMdd", null);
            DateTime today = DateTime.Today;
            int age = today.Year - birthDate.Year;
            if (today.Month < birthDate.Month || (today.Month == birthDate.Month && today.Day < birthDate.Day))
            {
                age--;
            }
            return age;
        }

        // 读取文件内容并返回患者列表
        public  List<Dictionary<string, string>> GetFileItems(string filePath)
        {
            List<Dictionary<string, string>> patientsList = new List<Dictionary<string, string>>();
            string[] lines = File.ReadAllLines(filePath);
            var patientsItem = new FixedSizeDeque<string>();
            foreach (var line in lines)
            {
                if (line.Contains("显示互认提示框")) continue;
                string patientName = "", patientId = "", patientBirthday = "", patientGender = "", patientDiagnose = "", patientNewItem = "";
                string patientPhone = "", patientAddress = "", doctorName = "", logDate = "", age = "";

                bool type1 = false, type2 = false;
                JObject entryInfo = null, leaveInfo = null;
                patientsItem.Add(line);
                if (line.Contains("要打开"))
                {
                    // 处理日志信息（模拟）
                    string[] separators = { "入参:" };
                    string[] secondSeparators = { "出参:" };
                    var entryLine = patientsItem.GetAll()[0];
                    var leaveLine = patientsItem.GetAll()[1];
                    logDate = entryLine.Split(new string[] { "====>" }, StringSplitOptions.None)[0];
                    // 解析 JSON 数据
                    try
                    {
                        entryInfo = JObject.Parse(entryLine.Split(separators, StringSplitOptions.None)[1]);
                        leaveInfo = JObject.Parse(leaveLine.Split(secondSeparators, StringSplitOptions.None)[1]);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine(ex.Message);

                    }
                    if (entryInfo == null || leaveInfo == null)
                    {
                        continue;
                    }

                    if (entryLine.Contains("患者近期相似检查项目查询接口"))
                    {
                        Console.WriteLine("患者近期相似检查项目查询接口");
                        type1 = true;

                        patientName = entryInfo["patience"]["name"].ToString();

                        patientId = entryInfo["patience"]["id_no"].ToString();
                        patientBirthday = entryInfo["patience"]["birthday"].ToString().Replace("-", "");

                        patientGender = entryInfo["patience"]["gender"].ToString() == "1" ? "男" : "女";

                        //patientDiagnose = query_from_local_db(patient_id, QueryType.DIAGNOSE)

                        patientPhone = entryInfo["patience"]["phone"].ToString();
                        patientAddress = entryInfo["patience"]["address"].ToString();
                        doctorName = entryInfo["doctor"]["name"].ToString();
                        var projectList = entryInfo["visit_order"]["project_list"];
                        var listNewItem = projectList
                                .Select(item => $"{item["project_name"]}:{item["study_body_part"]}")
                                .ToList();
                        patientNewItem = String.Join(" ", listNewItem);





                    }

                    if (leaveLine.Contains("您开立的项目有可互认的结果，请确认是否互认"))
                    {
                        type2 = true;
                        patientId = entryInfo["messages"]["PID"]["identityNo"].ToString();
                        patientName = entryInfo["messages"]["PID"]["personalName"].ToString();
                        patientGender = entryInfo["messages"]["PID"]["sexCode"].ToString() == "1" ? "男" : "女";
                        patientBirthday = entryInfo["messages"]["PID"]["birthDate"].ToString();
                        patientDiagnose = entryInfo["messages"]["DG1"]["diagName"].ToString();
                        doctorName = entryInfo["messages"]["PV1"]["doctorName"].ToString();
                        patientNewItem = entryInfo["messages"]["OBR"]["lab"][0]["labItemName"].ToString();


                        patientPhone = "";
                        patientAddress = "";

                    }

                    if (type1 || type2)
                    {
                        age = CalculateAge(patientBirthday).ToString();
                        Dictionary<string, string> patientItemToSave = new Dictionary<string, string>
                             {
                    { "patient_id", patientId },
                    { "name", patientName },
                    { "doctor", doctorName },
                    { "birthday", patientBirthday },
                    { "gender", patientGender },
                    { "age", $"{age}岁" },
                    { "diagnose", patientDiagnose },
                    { "patient_phone", patientPhone },
                    { "patient_address", patientAddress },
                    { "new_item", patientNewItem },
                    { "log_date", logDate }
                          };
                        patientsList.Add(patientItemToSave);
                    }

                }
            }

            return patientsList;
        }

        // 过滤文件并插入数据库
        public  void FilterFilesAndInsert(string directory, MySqlConnection connection)
        {
            var files = Directory.GetFiles(directory, "*.txt");
            foreach (var file in files)
            {
                var patientsList = GetFileItems(file);
                if (patientsList.Count == 0)
                { continue; }
                InsertDb(patientsList, connection);
            }
        }

        // 查询数据库是否存在记录
        public  bool QueryDb(string logDate, MySqlConnection connection)
        {
            string sql = $"SELECT * FROM recognition WHERE LogDate = @logDate";
            using (MySqlCommand cmd = new MySqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@logDate", logDate);
                var result = cmd.ExecuteScalar();
                return result != null;
            }
        }

        // 插入数据库
        public  void InsertDb(List<Dictionary<string, string>> patientsList, MySqlConnection connection)
        {
            string sql = @"INSERT INTO recognition 
                        (LogDate, Name, PatientID, Birthday, Gender, Age, Diagnose, Address, NewItem, Phone, Doctor) 
                        VALUES (@log_date, @name, @patient_id, @birthday, @gender, @age, @diagnose, @address, @new_item, @phone, @doctor)";

            foreach (var patient in patientsList)
            {
                if (QueryDb(patient["log_date"], connection))
                {
                    Console.WriteLine("Data already exists in recognition.");
                    continue;
                }

                using (MySqlCommand cmd = new MySqlCommand(sql, connection))
                {
                    cmd.Parameters.AddWithValue("@log_date", patient["log_date"]);
                    cmd.Parameters.AddWithValue("@name", patient["name"]);
                    cmd.Parameters.AddWithValue("@patient_id", patient["patient_id"]);
                    cmd.Parameters.AddWithValue("@birthday", patient["birthday"]);
                    cmd.Parameters.AddWithValue("@gender", patient["gender"]);
                    cmd.Parameters.AddWithValue("@age", patient["age"]);
                    cmd.Parameters.AddWithValue("@diagnose", patient["diagnose"]);
                    cmd.Parameters.AddWithValue("@address", patient["patient_address"]);
                    cmd.Parameters.AddWithValue("@new_item", string.Join(",", patient["new_item"])); // 假设是一个字符串数组
                    cmd.Parameters.AddWithValue("@phone", patient["patient_phone"]);
                    cmd.Parameters.AddWithValue("@doctor", patient["doctor"]);

                    cmd.ExecuteNonQuery();
                    Console.WriteLine("Data inserted successfully!");
                }
            }
        }

        // 创建数据库
        public static void CreateDb(MySqlConnection connection)
        {
            string createTableSql = @"
            CREATE TABLE IF NOT EXISTS recognition (
                LogDate VARCHAR(255)  PRIMARY KEY,
                Name VARCHAR(255),
                PatientID VARCHAR(255),
                Birthday VARCHAR(255),
                Gender VARCHAR(255),
                Age VARCHAR(255),
                Diagnose VARCHAR(255),
                Address VARCHAR(255),
                NewItem VARCHAR(255),
                Phone VARCHAR(255),
                Doctor VARCHAR(255)
            );
        ";

            using (MySqlCommand cmd = new MySqlCommand(createTableSql, connection))
            {
                cmd.ExecuteNonQuery();
            }

            string createTableSql2 = @"
            CREATE TABLE IF NOT EXISTS zlbhPath (
                directory VARCHAR(255) PRIMARY KEY
            );
        ";

            using (MySqlCommand cmd = new MySqlCommand(createTableSql2, connection))
            {
                cmd.ExecuteNonQuery();
            }
        }

        // 主函数
        public void  Start()
        {

            MySqlConnection conn = new MySqlConnection(connectionString);

            conn.Open();
            CreateDb(conn);
            FilterFilesAndInsert(directory, conn);
            conn.Close();
        }

        public class FixedSizeDeque<T>
        {
            private LinkedList<T> deque;
            private int maxSize;

            public FixedSizeDeque(int maxSize = 3)
            {
                this.maxSize = maxSize;
                this.deque = new LinkedList<T>();
            }

            public void Add(T item)
            {
                if (deque.Count == maxSize)
                {
                    deque.RemoveFirst(); // 删除队列的第一个元素
                }
                deque.AddLast(item); // 添加新元素
            }

            public List<T> GetAll()
            {
                return new List<T>(deque); // 返回队列中的所有元素
            }
        }
    }
}
